### PR TITLE
GFORMS-845 - Display width changes

### PR DESCRIPTION
**Pull request by SandipKhedekar**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 1]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+  val defaultDisplayWidthLSizeForText = 20
+  val defaultDisplayWidthXlSizeForText = 30
+  val defaultDisplayWidthXxlSizeForText = 40
+  val defaultDisplayWidthXsSizeForNumber = 2
+  val defaultDisplayWidthSSizeForNumber = 3
+  val defaultDisplayWidthMSizeForNumber = 4
+  val defaultDisplayWidthLSizeForNumber = 5
+  val defaultDisplayWidthXlSizeForNumber = 10
+  val defaultDisplayWidthXxlSizeForNumber = 20
+  val defaultSizeNumber = 10
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:298]: This seems not to be used

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 2]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+  val defaultSizePositiveNumber = 10
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:299]: This seems not to be used

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 3]
```diff
@@ -160,7 +160,81 @@ object RoundingMode {
   )
 }
 
-sealed trait TextConstraint
+sealed trait TextConstraint {
+  val defaultSize: Int = this match {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:164]: Can we make this more readable like this:

```
sealed trait TextConstraint {
  val defaultSize: Int = this match {
    case Number(maxWholeDigits, maxFractionalDigits, _, _) =>
      if (maxFractionalDigits > 0)
        maxWholeDigits + maxFractionalDigits + 1
      else
        maxWholeDigits
    case PositiveNumber(maxWholeDigits, maxFractionalDigits, _, _) =>
      if (maxFractionalDigits > 0)
        maxWholeDigits + maxFractionalDigits + 1
      else
        maxWholeDigits
    case BasicText                    => TextConstraint.defaultSizeBasicText
    case ShortText(_, max)            => max
    case Lookup(_)                    => TextConstraint.defaultSizeLookup
    case TextWithRestrictions(_, max) => max
    case Sterling(_, _)               => TextConstraint.defaultSizeSterling
    case UkBankAccountNumber          => TextConstraint.defaultSizeUkBankAccountNumber
    case UkSortCodeFormat             => TextConstraint.defaultSizeUkSortCodeFormat
    case SubmissionRefFormat          => TextConstraint.defaultSizeSubmissionRefFormat
    case TelephoneNumber              => TextConstraint.defaultSizeTelephoneNumber
    case Email                        => TextConstraint.defaultSizeEmail
    case EmailVerifiedBy(_, _)        => TextConstraint.defaultSizeEmailVerifiedBy
    case UTR                          => TextConstraint.defaultSizeUTR
    case NINO                         => TextConstraint.defaultSizeNINO
    case UkVrn                        => TextConstraint.defaultSizeUkVrn
    case CountryCode                  => TextConstraint.defaultSizeCountryCode
    case NonUkCountryCode             => TextConstraint.defaultSizeNonUkCountryCode
    case CompanyRegistrationNumber    => TextConstraint.defaultSizeCompanyRegistrationNumber
    case EORI                         => TextConstraint.defaultSizeEORI
    case UkEORI                       => TextConstraint.defaultSizeUkEORI
    case ChildBenefitNumber           => TextConstraint.defaultSizeChildBenefitNumber
  }
}
```

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 4]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+  }
+
+  def getSizeForDisplayWidthForNumber(displayWidth: DisplayWidth.DisplayWidth): Int = displayWidth match {
+    case DisplayWidth.XS  => TextConstraint.defaultDisplayWidthXsSizeForNumber
+    case DisplayWidth.S   => TextConstraint.defaultDisplayWidthSSizeForNumber
+    case DisplayWidth.M   => TextConstraint.defaultDisplayWidthMSizeForNumber
+    case DisplayWidth.L   => TextConstraint.defaultDisplayWidthLSizeForNumber
+    case DisplayWidth.XL  => TextConstraint.defaultDisplayWidthXlSizeForNumber
+    case DisplayWidth.XXL => TextConstraint.defaultDisplayWidthXxlSizeForNumber
+  }
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:341]: Just out of curiosity, why not to do simple:
```
  def getSizeForDisplayWidthForText(displayWidth: DisplayWidth.DisplayWidth): Int = displayWidth match {
    case DisplayWidth.XS  => 4
    case DisplayWidth.S   => 5
    case DisplayWidth.M   => 10
    case DisplayWidth.L   => 20
    case DisplayWidth.XL  => 30
    case DisplayWidth.XXL => 40
  }

  def getSizeForDisplayWidthForNumber(displayWidth: DisplayWidth.DisplayWidth): Int = displayWidth match {
    case DisplayWidth.XS  => 2
    case DisplayWidth.S   => 3
    case DisplayWidth.M   => 4
    case DisplayWidth.L   => 5
    case DisplayWidth.XL  => 10
    case DisplayWidth.XXL => 20
  }
```
?

It is more readable (imho) and much less typing.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 5]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+    constraint match {
+      case Number(_, _, _, _) | PositiveNumber(_, _, _, _) | Sterling(_, _) | UkBankAccountNumber | UkSortCodeFormat =>
+        getSizeClassForNumbers(constraint, displayWidth)
+
+      case _ =>
+        getSizeClassForText(constraint, displayWidth)
+    }
+
+  private def getSizeClassForText(constraint: TextConstraint, displayWidth: DisplayWidth.DisplayWidth): String = {
+    val size = (constraint, displayWidth) match {
```
[the remark this answers is not in the document]
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:353]: No need to match `constraint` as `constraint.defaultSize` will give us correct value.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 6]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+      case 5                      => "govuk-input--width-5"
+      case n if n > 5 && n <= 10  => "govuk-input--width-10"
+      case n if n > 10 && n <= 20 => "govuk-input--width-20"
+      case n if n > 20 && n <= 30 => "govuk-input--width-30"
+      case n if n > 30            => "govuk-input--width-40"
+    }
+  }
+
+  private def getSizeClassForNumbers(constraint: TextConstraint, displayWidth: DisplayWidth.DisplayWidth): String = {
+    val size = (constraint, displayWidth) match {
```
[the remark this answers is not in the document]
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:374]: Same as above.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 7]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+      case Number(_, _, _, _) | PositiveNumber(_, _, _, _) | Sterling(_, _) | UkBankAccountNumber | UkSortCodeFormat =>
+        getSizeClassForNumbers(constraint, displayWidth)
+
+      case _ =>
+        getSizeClassForText(constraint, displayWidth)
+    }
+
+  private def getSizeClassForText(constraint: TextConstraint, displayWidth: DisplayWidth.DisplayWidth): String = {
+    val size = (constraint, displayWidth) match {
+      case (_, displayWidth) if displayWidth != DisplayWidth.DEFAULT =>
```
[the remark this answers is not in the document]
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:354]: Correct.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 8]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
...
+      case n if n > 5 && n <= 10  => "govuk-input--width-10"
+      case n if n > 10 && n <= 20 => "govuk-input--width-20"
+      case n if n > 20 && n <= 30 => "govuk-input--width-30"
+      case n if n > 30            => "govuk-input--width-40"
+    }
+  }
+
+  private def getSizeClassForNumbers(constraint: TextConstraint, displayWidth: DisplayWidth.DisplayWidth): String = {
+    val size = (constraint, displayWidth) match {
+      case (_, displayWidth) if displayWidth != DisplayWidth.DEFAULT =>
```
[the remark this answers is not in the document]
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:375]: Correct.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 9]
```diff
@@ -208,10 +282,114 @@ case object ChildBenefitNumber extends TextConstraint
 object TextConstraint {
   val defaultWholeDigits = 11
   val defaultFactionalDigits = 2
+  val defaultDisplayWidthSize = 20
```
[the remark this answers is not in the document]
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:285]: Yes. Similarly `defaultSizeNumber`, `defaultSizePositiveNumber`, `defaultSizeShortText`, `defaultSizeTextWithRestrictions`. I just kept them for reference or may be used in future. We can remove them.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala (hunk 1 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:298]: Ok. I believe you are talking about `defaultSizeNumber` and `defaultSizePositiveNumber`. I mentioned in the above comment.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala (hunk 2 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:299]: Same as above.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala (hunk 3 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:164]: Agreed.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala (hunk 4 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:341]: I wanted to define them as constants. At the moment, these constants are used at only one place. In future, they may be used in other places. And if any changes in them will be reflected in all used places and avoid defects.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 10]
```diff
@@ -160,7 +160,67 @@ object RoundingMode {
...
+        else
+          maxWholeDigits
+
+      deriveCssClassNameForNumber(size)
+    case PositiveNumber(maxWholeDigits, maxFractionalDigits, _, _) =>
+      val size =
+        if (maxFractionalDigits > 0)
+          maxWholeDigits + maxFractionalDigits + 1
+        else
+          maxWholeDigits
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:178]: Let's push this logic down to get rid of duplication from lines 166-170.

(if appropriate length identifiers would be used whole `match` expression will became nicely aligned)
```
case Number(whole, fractional, _, _)         => deriveCssClassNameForNumber(whole, fractional)
case PositiveNumber(whole, fractional, _, _) => deriveCssClassNameForNumber(whole, fractional)
case BasicText                               => CssClassSize._20
...
```

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 11]
```diff
@@ -160,7 +160,67 @@ object RoundingMode {
...
+    case UkVrn                        => CssClassSize._10
+    case CountryCode                  => CssClassSize._5
+    case NonUkCountryCode             => CssClassSize._5
+    case CompanyRegistrationNumber    => CssClassSize._20
+    case EORI                         => CssClassSize._20
+    case UkEORI                       => CssClassSize._20
+    case ChildBenefitNumber           => CssClassSize._20
+  }
+
+  def deriveCssClassNameForText(n: Int): String = n match {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:203]: can be `private`

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 12]
```diff
@@ -160,7 +160,67 @@ object RoundingMode {
...
+    case 3                      => CssClassSize._3
+    case 4                      => CssClassSize._4
+    case 5                      => CssClassSize._5
+    case n if n > 5 && n <= 10  => CssClassSize._10
+    case n if n > 10 && n <= 20 => CssClassSize._20
+    case n if n > 20 && n <= 30 => CssClassSize._30
+    case n if n > 30            => CssClassSize._40
+  }
+
+  def deriveCssClassNameForNumber(n: Int): String = n match {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:214]: can be `private`

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 13]
```diff
@@ -212,6 +272,49 @@ object TextConstraint {
   implicit val format: OFormat[TextConstraint] = derived.oformat[TextConstraint]
 
   def filterNumberValue(s: String): String = s.filterNot(c => (c == '£'))
+
+  def getSizeForDisplayWidthForText(displayWidth: DisplayWidth.DisplayWidth): String = displayWidth match {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:276]: can be `private`

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala [hunk 14]
```diff
@@ -212,6 +272,49 @@ object TextConstraint {
...
+    case DisplayWidth.XS  => CssClassSize._4
+    case DisplayWidth.S   => CssClassSize._5
+    case DisplayWidth.M   => CssClassSize._10
+    case DisplayWidth.L   => CssClassSize._20
+    case DisplayWidth.XL  => CssClassSize._30
+    case DisplayWidth.XXL => CssClassSize._40
+  }
+
+  def getSizeForDisplayWidthForNumber(displayWidth: DisplayWidth.DisplayWidth): String = displayWidth match {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala:285]: can be `private`

**Pull request merged**